test/lit/passes/simplify-globals-read_only_to_write.wast

```diff
@@ -386,4 +352,187 @@
       (i32.const 1)
     )
   )
+
+  ;; CHECK:      (func $foo (param $x i32) (result i32)
+  ;; CHECK-NEXT:  (unreachable)
+  ;; CHECK-NEXT: )
+  (func $foo (param $x i32) (result i32)
+    (unreachable)
+  )
+)
+
+;; Using the global's value in a way that can cause side effects prevents the
+;; read-only-to-write optimization.
+(module
+  ;; CHECK:      (type $none_=>_none (func))
+
+  ;; CHECK:      (type $none_=>_i32 (func (result i32)))
+
+  ;; CHECK:      (global $global (mut i32) (i32.const 0))
+  (global $global (mut i32) (i32.const 0))
+  ;; CHECK:      (global $other i32 (i32.const 0))
+  (global $other (mut i32) (i32.const 0))
+  ;; CHECK:      (func $side-effects-in-condition
+  ;; CHECK-NEXT:  (if
+  ;; CHECK-NEXT:   (if (result i32)
+  ;; CHECK-NEXT:    (global.get $global)
+  ;; CHECK-NEXT:    (call $foo)
+  ;; CHECK-NEXT:    (i32.const 1)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (global.set $global
+  ;; CHECK-NEXT:    (i32.const 1)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $side-effects-in-condition
+    (if
+      (if (result i32)
+        (global.get $global) ;; the global's value may cause foo() to be called
+        (call $foo)
+        (i32.const 1)
+      )
+      (global.set $global (i32.const 1))
+    )
+  )
+
+  ;; CHECK:      (func $foo (result i32)
+  ;; CHECK-NEXT:  (unreachable)
+  ;; CHECK-NEXT: )
+  (func $foo (result i32)
+    (unreachable)
+  )
+)
+
+;; As above, but now the global's value is not the condition of the if, so there
+;; is no problem.
+(module
+  ;; CHECK:      (type $none_=>_none (func))
+
+  ;; CHECK:      (type $none_=>_i32 (func (result i32)))
+
+  ;; CHECK:      (global $global i32 (i32.const 0))
+  (global $global (mut i32) (i32.const 0))
+  ;; CHECK:      (global $other i32 (i32.const 0))
+  (global $other (mut i32) (i32.const 0))
+  ;; CHECK:      (func $side-effects-in-condition-2
+  ;; CHECK-NEXT:  (if
+  ;; CHECK-NEXT:   (if (result i32)
+  ;; CHECK-NEXT:    (call $foo)
+  ;; CHECK-NEXT:    (i32.const 1)
+  ;; CHECK-NEXT:    (i32.const 0)
```

Comment:
This `i32.const 0` is chosen based on the global's initial value, right?

Reply:
Yes. The other parts of this pass recognize that that global has no more writes, so it is immutable. And then we propagate immutable global constants.